% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coverage_matrix_bwtool.R
\name{coverage_matrix_bwtool}
\alias{coverage_matrix_bwtool}
\title{Given a set of regions for a chromosome, compute the coverage matrix for a
given SRA study using \code{bwtool} as a 
\link[SummarizedExperiment]{RangedSummarizedExperiment-class} object.}
\usage{
coverage_matrix_bwtool(project, regions,
  bwtool = "/dcl01/leek/data/bwtool/bwtool-1.0/bwtool", bpparam = NULL,
  outdir = NULL, verbose = TRUE, sumsdir = tempdir(), bed = NULL,
  url_table = NULL, commands_only = FALSE, pheno = NULL, ...)
}
\arguments{
\item{project}{A character vector with one SRA study id.}

\item{regions}{A \link[GenomicRanges]{GRanges-class} object with regions
for which to calculate the coverage matrix.}

\item{bwtool}{The path to \code{bwtool}. Uses as the default the
location at JHPCE.}

\item{bpparam}{A \link[BiocParallel]{BiocParallelParam-class} instance which
will be used to calculate the coverage matrix in parallel. By default, 
\link[BiocParallel]{SerialParam-class} will be used.}

\item{outdir}{The destination directory for the downloaded file(s) that were
previously downloaded with \link{download_study}. If the files are missing, 
but \code{outdir} is specified, they will get downloaded first. By default
\code{outdir} is set to \code{NULL} which will use the data from the web.
We only recommend downloading the full data if you will use it several times.
Note that if you are working at JHPCE or SciServer, the files will be
located automatically.}

\item{verbose}{If \code{TRUE} basic status updates will be printed along the 
way.}

\item{sumsdir}{The path to an existing directory where the \code{bwtool}
sum tsv files will be saved.}

\item{bed}{The path to the BED file for the regions. You are responsible
for making sure that the BED file and the regions are in the same order.
Could be useful for a scenario where you have a BED file and import it to
define \code{regions}.}

\item{url_table}{A custom data.frame named with the same columns as
\code{recount::recount_url}. If \code{NULL}, the default is 
\code{recount::recount_url}.
Use \code{local_url}
saved in \code{/dcl01/leek/data/recount-website/fileinfo/local_url.RData}.}

\item{commands_only}{If \code{TRUE} the bwtool commands will be saved in a
file called recount-bwtool-commands_PROJECT.txt and exit without running
\code{bwtool}. This is useful if you have a very large regions set and want
to run the commands in an array job. Then run
\code{coverage_matrix_bwtool(commands_only = FALSE)} to create the RSE
object(s).}

\item{pheno}{\code{NULL} by default. Specify only if you are using a custom
metadata table.}

\item{...}{Additional arguments passed to \link{download_study} when
\code{outdir} is specified but the required files are missing.}
}
\value{
A \link[SummarizedExperiment]{RangedSummarizedExperiment-class}
object with the counts stored in the assays slot.
}
\description{
Given a set of genomic regions as created by 
\link[recount]{expressed_regions}, this
function computes the coverage matrix using \code{bwtool}. You can
then scale the counts to a 40 million 100 bp reads library using
\link[recount]{scale_counts}.
}
\details{
When using \code{outdir = NULL} the information will be accessed
from the web on the fly. If you encounter internet access problems, it might
be best to first download the BigWig files using \link{download_study}. This
might be the best option if you are accessing all chromosomes for a given
project and/or are thinking of using different sets of \code{regions} (for
example, from different cutoffs applied to \link{expressed_regions}).
If you are working at JHPCE (and part of \code{leekgroup}) or at SciServer,
the files will be located automatically.

Check also \code{system.file('extdata', 'jhpce', package = 'recount.bwtool')}
for some scripts that will run this function for all the projects we have
available at JHPCE.
}
\examples{

if(.Platform$OS.type != 'windows') {
## Disable the example for now. I'd have to figure out how to install
## bwtool on travis
if(FALSE) {
    ## Reading BigWig files is not supported by rtracklayer on Windows
    ## (only needed for defining the regions in this example)
    ## Define expressed regions for study DRP002835, chrY
    regions <- expressed_regions('DRP002835', 'chrY', cutoff = 5L, 
        maxClusterGap = 3000L)

    ## Now calculate the coverage matrix for this study
    rse <- coverage_matrix_bwtool('DRP002835', regions)

    ## Scale counts
    rse_scaled <- scale_counts(rse, round = FALSE)
}
}

}
\seealso{
\link[recount]{coverage_matrix}, \link[recount]{download_study},
\link[derfinder]{findRegions}, \link[derfinder]{railMatrix}
}
\author{
Leonardo Collado-Torres
}
